feat(pg): add Cloud SQL + IAM authentication support

nuritizra opened the pull request:
## Description

Add support for Google Cloud SQL + IAM by allowing a raw `pg.ClientConfig` to be passed to `PostgresStore`.

Currently, `PostgresConfig` only accepts either:

* a connection string, or
* a static TCP config (`{host, port, database, user, password, ssl}`).

This breaks when using the Cloud SQL Node.js Connector with IAM auth, because it doesn’t expose a resolvable host.

When Backstage is configured with:

```yaml
database:
  client: pg
  connection:
    type: cloudsql
    instance: ${DATABASE_INSTANCE}
    user: ${DATABASE_USER}
    ipAddressType: PRIVATE
```

the Cloud SQL connector is used ([[see `postgres.ts` in Backstage](https://github.com/backstage/backstage/blob/master/packages/backend-defaults/src/entrypoints/database/connectors/postgres.ts#L111)]
That connector establishes a secure stream and returns a `pg.ClientConfig` with:

* `stream`: a TLS socket connected to Cloud SQL
* `ssl`: TLS options
* `user`, `database`
* `password`: often a **function** that dynamically generates IAM tokens

There is **no stable DNS host** to extract. Forcing `{host, port}` is incorrect and fails at runtime.
`pg` and `pg-promise` already support this shape — we just need to accept and forward it.

### Changes in this PR

* Extend `PostgresConfig` to accept a raw `pg.ClientConfig`.
* Relax constructor validation to allow configs with `stream` or a functional `password`.
* Forward connector configs directly to `pg-promise`.

This preserves full backward compatibility with connection strings and TCP configs.

## Type of Change

* [ ] Bug fix (non-breaking change that fixes an issue)
* [x] New feature (non-breaking change that adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Documentation update
* [ ] Code refactoring
* [ ] Performance improvement
* [ ] Test update

---

## Checklist

* [ ] I have made corresponding changes to the documentation (if applicable)
* [x] I have added tests that prove my fix is effective or that my feature works

nuritizra committed:
feat(pg): add Cloud SQL + IAM authentication support

- Extend PostgresConfig to accept pg.ClientConfig for Cloud SQL connector
- Add validation for stream-based and password function configs
- Maintain backward compatibility with existing connection patterns
- Add comprehensive tests for all three config types:
  * Connection string configs
  * TCP host/port configs
  * Cloud SQL connector configs (stream + IAM auth)

Fixes integration with Google Cloud SQL Node.js Connector when using
IAM authentication, which provides configs with stream sockets and
dynamic password functions instead of traditional host/port.

NikAiyer commented on stores/pg/src/storage/index.ts:
```diff
@@ -79,23 +88,45 @@ export class PostgresStore extends MastraStorage {
+          idleTimeoutMillis: (config as any).idleTimeoutMillis ?? undefined,
+        } as any;
+      } else if ('host' in config) {
+        this.#config = {
+          host: (config as any).host,
+          port: (config as any).port,
+          database: (config as any).database,
+          user: (config as any).user,
+          password: (config as any).password,
+          ssl: (config as any).ssl,
+          max: config.max,
+          idleTimeoutMillis: config.idleTimeoutMillis,
```
That way they're more specifically typed and we won't need to rely on using assertions as much.

nuritizra committed:
refactor(pg): improve type safety and fix linting issues

- Add type guards for PostgresConfig variants to eliminate unsafe 'as any' casts
- Replace manual type checking with proper type narrowing functions
- Remove redundant nullish coalescing with undefined assignments
- Fix import order: move pg import before pg-promise import
- Enhance validation logic with better type safety throughout

Addresses PR feedback for improved type safety and code quality while
maintaining all existing functionality and test coverage.

nuritizra committed:
Merge branch 'main' into feat/postgres-cloud-sql-support

nuritizra committed:
Merge branch 'main' into feat/postgres-cloud-sql-support

nuritizra committed:
Merge branch 'main' into feat/postgres-cloud-sql-support

NikAiyer committed:
Merge branch 'main' into feat/postgres-cloud-sql-support

NikAiyer reviewed (approved):
LGTM! Thanks for making this PR!
